Automattic/newspack-plugin
# feat: activate and deactivate are bulk publish controls; add unpublished option for previewing

Instead of using an "active" group, campaigns are displayed on the front-end if they are published and if they belong to a segment matching the current reader.

### All Submissions:






### Changes proposed in this Pull Request:



Requires `newspack-popups` be on `feat/current-campaign-group-pub-status`, see: https://github.com/Automattic/newspack-popups/pull/375

Like in #775, introduces the concept of bulk-activating and deactivating campaigns by group. Unlike #775, this branch uses the WP post status of campaigns to determine whether or not individual campaigns should be shown on the front-end, and turns "Activate" and "Deactivate" buttons into bulk publish/unpublish actions that apply to the currently viewed campaigns. "Activating" campaigns when viewing a particular group of campaigns will publish any unpublished campaigns in that group and also automatically unpublish all campaigns that don't belong to the viewed group. This results in an editor experience very similar to that of #775, but without requiring the additional "active group" status for campaigns to appear on the front-end. This also avoids an expensive tax query to fetch campaigns on the front-end, instead relying on post IDs only.

Also adds a new option in the preview UIs to let users choose whether to show unpublished campaigns. If checked, the preview modal will show draft, pending, scheduled and published campaigns. If unchecked, only published campaigns will be shown (matching the front-end display).

<img width="270" alt="Screen Shot 2021-01-08 at 3 12 05 PM" src="https://user-images.githubusercontent.com/2230142/104072269-23bf9800-51c8-11eb-9cbb-014197005392.png">

~Note that for test purposes I've removed labels for Test and "Never" frequency campaigns and ignored those criteria when displaying campaigns in anticipation of #776, but I did not fully deprecate all functionality for those statuses here.~ I've restored these for now so that we can more cleanly deprecate them in the other PR.

### How to test the changes in this Pull Request:

1. View the Campaigns tab in the wizard. Confirm that the filter always defaults to the "All campaigns" filter view.
2. Confirm that published campaigns are shown with a green label and unpublished campaigns (draft, pending or scheduled) are shown with a grey label.
3. Expand the popover menu for a published campaign. Confirm that there's a new "Unpublish" option that when clicked, reverts the selected campaign to draft status.
4. Observe the "Publish All", "Unpublish All" and "Preview" buttons that are now visible in the All Campaigns filter view.
5. Click "Publish All" and confirm that this sets ALL campaigns to "publish" status. Then click "Unpublish All" and confirm that it sets ALL campaigns to "draft" status.
6. Filter the campaigns list by a group. Click "Publish All" and confirm that this sets only the campaigns in this group to "publish" status. (Make sure to view another group to confirm that those campaigns are still in draft status).
7. View another group and click "Publish All". Confirm that the campaigns in this group are now set to "publish" status, and campaigns in the first group that don't exist in this second group are set to "draft" status.
8. Publish some campaigns from various groups. Then filter by one group and click "Unpublish All." Confirm that only the campaigns in the filtered group are unpublished.
9. In the preview popover, observe a new checkbox option to "Show unpublished campaigns". Preview various groups and segments and confirm that preview still works as expected. Also confirm that draft, pending, scheduled and published campaigns are all shown in the preview modal if the "Show unpublished campaigns" option is selected. Repeat these tests with both the Campaigns and Preview tabs.
10. View the site front-end in a new incognito window and confirm that only published campaigns matching the reader's segment are shown.
11. The one caveat to (10) is that campaigns that are set to "Uncategorized" are now able to show with any category—this seems more intuitive to me, especially since uncategorized campaigns are automatically assigned the "Uncategorized" category when published.

### Other information:
